### ISSUE TITLE
distances between pairs of wavelet packets of images

### ISSUE THREAD

**Issue by antoineszatkownik** [BonnBytes/PyTorch-FWD#30]
Hello,

Thank you for this very nice work,

Your python library is designed to compute a single float score for a whole dataset, but in the paper, Fig. 1. you show MSE between two wavelet packets for two different images. 
It seems doable by tweaking a little bit the `compute_packet_statistics()` function in `fwd.py` so that it returns `packet_tensor` of shape `(P, N, C * H * W)`, where `N` is dataset size.
I guess you then compute the MSE between flattened tensors of shape `(1, P*C*H*W)`, right?

This is how I modified the function (ie it just returns `packet_tensor` and not the means and covariances)

`def compute_packet_statistics(
    dataloader, wavelet, max_level, log_scale
):
    """Compute wavelet packet transform across batches.

    Args:
        dataloader (th.utils.data.DataLoader): Torch dataloader.
        wavelet (str): Choice of wavelet.
        max_level (int): Wavelet decomposition level.
        log_scale (bool): Apply log scale.

    Returns:
        Tuple[np.ndarray, ...]: Mean and sigma for each packet.
    """
    packets = []
    device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")
    for img_batch in tqdm(dataloader):
        if isinstance(img_batch, list):
            img_batch = img_batch[0]
        img_batch = img_batch.to(device)
        packets.append(
            forward_wavelet_packet_transform(
                img_batch, wavelet, max_level, log_scale
            ).cpu()
        )
    packet_tensor = torch.cat(packets, dim=0)
    print(packet_tensor.shape)
    packet_tensor = torch.permute(packet_tensor, (1, 0, 2, 3, 4))
    P, BS, C, H, W = packet_tensor.shape
    packet_tensor = torch.reshape(packet_tensor, (P, BS, C * H * W))

    return packet_tensor

wavelet = "sym5"
max_level = 4
log_scale = False`

I tried to apply that to CIFAR10 images (which are `torch.Size([N, 3, 32, 32])`)

`train_wavelet = compute_packet_statistics(train_loader, wavelet, max_level, log_scale)
`
returns a tensor of shape `torch.Size([50000, 256, 3, 10, 10])`

I just want to confirm with you whether it's correct for me to compute distances/MSE on these flattened vectors which will have` 256*3*10*10 = 76800` wavelet features? If I want to compute all pairwise distances, this leads to high computational time...

Thanks in advance

**Comment by loki-veera** [comment]
Hello @antoineszatkownik,

Thank you for the question. 

> I guess you then compute the MSE between flattened tensors of shape (1, P*C*H*W), right?

Yes, we compute the MSE score for the flattened tensor. 

>  whether it's correct for me to compute distances/MSE on these flattened vectors which will have 256310*10 = 76800 wavelet features?

Regarding this question, firstly we use 'Haar' wavelet instead of 'sym5'. Thanks for raising this point, we will make Haar default in next release ASAP.
 
Secondly, as stated in the paper Section:4.4, we use level 4 packet transform for 256x256 image size and further reduce the packet transform level by 1 for every half image size. I.e., for CIFAR10, the packet transform should be level 1.

Factoring in these new parameters, the resultant shape of tensor from your modified `compute_packet_statistics` function should be `torch.Size([50000, 4, 3, 16, 16])`.
Further flattening would result only in `3072` features per image. 

I hope this answers your question.

**Comment by antoineszatkownik** [comment]
Yes thank you for your answer.

I have another question, why do you use the Haar wavelet?

Moreover, I guess this somehow has to be expected, but when I use the Haar wavelet with level 1 and compute pairwise euclidean distances on the packets produced for CIFAR10 (or whatever data), then the distances behave as re-scaled euclidean distances computed directly in pixel space. Do you know why?

Thank you

**Comment by loki-veera** [comment]
We use Haar primarily due to computational efficiency, which is beneficial as image resolution increases. Additionally, the user-study presented in our paper suggests it performs well in practice when evaluated with Stable Diffusion models.

Yes this behaviour is expected. Given the smaller image size and use of only level 1 packet transform, much of the spatial structure in images is retained. To extract more abstract and high level features, we recommend applying deeper levels of packet transform in your use-case.

**Comment by loki-veera** [comment]
I hope all your questions are answered, and I am closing this issue.
